File: server/src/main/java/io/microconfig/server/vault/VaultClientImpl.java
package io.microconfig.server.vault;

import com.fasterxml.jackson.databind.JsonNode;
import io.microconfig.server.vault.credentials.VaultCredentials;
import io.microconfig.server.vault.exceptions.VaultAuthException;
import io.microconfig.server.vault.exceptions.VaultSecretNotFound;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.net.URI;
import java.net.http.HttpRequest;
import java.time.Duration;

import static io.microconfig.server.util.HttpUtil.httpSend;
import static io.microconfig.server.util.JsonUtil.parseJson;

@Slf4j
@RequiredArgsConstructor
@Service
public class VaultClientImpl implements VaultClient {
    private final VaultConfig config;

    @Override
    public String fetchSecret(VaultCredentials credentials, String property) {
        int dotIndex = property.lastIndexOf('.');
        String path = property.substring(0, dotIndex);
        String key = property.substring(dotIndex + 1);
        log.debug("Fetching {} {}", path, key);

        var node = readPath(path, credentials.getToken(config));
        var value = node.path("data").path("data").get(key);
        if (value == null) throw new VaultSecretNotFound(property);
        return value.asText();
    }

    private JsonNode readPath(String path, String token) {
        var splitPath = splitPath(path);
        var request = HttpRequest.newBuilder(URI.create(config.getAddress() + "/v1/" + splitPath[0] + "/data" + splitPath[1]))
            .setHeader("X-Vault-Token", token)
            .timeout(Duration.ofSeconds(2))
            .build();
        var response = httpSend(request);
        log.debug("Vault response {}", response.body());
        var node = parseJson(response.body());
        if (node.get("errors") != null) {
            throw new VaultAuthException();
        }
        return node;
    }

    private String[] splitPath(String path) {
        int slash = path.indexOf('/');
        return new String[]{path.substring(0, slash), path.substring(slash)};
    }
}

File: server/src/main/kotlin/io/microconfig/server/vault/VaultClientImpl.kt
package io.microconfig.server.vault

import com.fasterxml.jackson.databind.JsonNode
import io.microconfig.server.common.GET
import io.microconfig.server.common.httpClient
import io.microconfig.server.common.logger
import io.microconfig.server.common.send
import io.microconfig.server.vault.exceptions.VaultSecretNotFound

class VaultClientImpl(private val config: VaultConfig) : VaultClient {
    private val log = logger()
    private val http = httpClient()

    override fun fetchKV(property: String): String {
        val (path, key) = pathKey(property)
        log.debug("Fetching $path $key")
        val node = getPath(path)
        val value = node
            .path("data")
            .path("data")[key] ?: throw VaultSecretNotFound(property)
        return value.asText()
    }

    private fun pathKey(property: String): Pair<String, String> {
        val dotIndex = property.lastIndexOf('.')
        val path = property.substring(0, dotIndex)
        val key = property.substring(dotIndex + 1)
        return Pair(path, key)
    }

    private fun getPath(path: String): JsonNode {
        return GET(kvUrl(path))
            .build(config)
            .send(http)
            .vaultResponse()
    }

    private fun kvUrl(path: String): String {
        val (root, key) = splitPath(path)
        return "${config.address}/v1/$root/data/$key"
    }

    private fun splitPath(path: String): Pair<String, String> {
        val slash = path.indexOf('/')
        val root = path.substring(0, slash) // secret engine path like /secret
        val key = path.substring(slash + 1) // actual key path like dev/database
        return Pair(root, key)
    }
}

